### PR TITLE
Starting work for CollectionAssert

### PR DESCRIPTION
**Pull request by HaydenMeloche**
#### Check List:
* Fixes 2102
* Unit tests : not yet
* Javadoc with a code example (on API only) : not yet

Hoping I could get some initial overall feedback before I go to deep down the rabbit hole as well as help with my comment here:
https://github.com/assertj/assertj-core/issues/2102#issuecomment-767285633 @scordio 

Note: This PR is still in a rough state, I have not done any sort of documentation, unused imports are leftover etc. This will all be cleaned up by the time this PR is put in a ready state

### CODE REVIEW

**Comment by scordio** [comment]
Thank you @HaydenMeloche, I'll look into it today.

**Comment by scordio** [comment]
> hey @scordio got a chance to working on this over the weekend. I setup my assertion class as such
> 
> ```java
> public abstract class AbstractCollectionAssert <SELF extends AbstractCollectionAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>,
>   ACTUAL extends java.util.Collection<? extends ELEMENT>,
>   ELEMENT,
>   ELEMENT_ASSERT extends AbstractAssert<ELEMENT_ASSERT, ELEMENT>>
>   extends AbstractIterableAssert<SELF, ACTUAL, ELEMENT, ELEMENT_ASSERT>
>   implements IndexedObjectEnumerableAssert<SELF, ELEMENT> {
> ```
> 
> I'm running into an issue now with this setup. Existing test cases that test the `Iterable` type are now looking for a Collection assert rather than a Iterator assert.
> 
> Ex: This test case:
> 
> https://github.com/assertj/assertj-core/blob/4fad9a03993e66fd4e2735352c22c52d206e9a1e/src/test/java/org/assertj/core/api/Assertions_assertThat_with_Iterable_Test.java#L30
> 
> Gives error:
> 
> ```
> Required type:
> AbstractIterableAssert
> <?,
> java.lang.Iterable<? extends java.lang.Object>,
> java.lang.Object,
> org.assertj.core.api.ObjectAssert<java.lang.Object>>
> Provided:
> CollectionAssert
> <Object,
> 
> 
> 
> no instance(s) of type variable(s) ELEMENT, T exist so that Collection<? extends ELEMENT> conforms to Iterable<?>
> ```
> 
> In [Assertions.java](https://github.com/assertj/assertj-core/blob/main/src/main/java/org/assertj/core/api/Assertions.java) I setup the collection assert as
> 
> ```java
>   public static <ELEMENT> CollectionAssert<ELEMENT> assertThat(Collection<? extends ELEMENT> actual) {
>     return AssertionsForInterfaceTypes.assertThat(actual);
>   }
> ```
> 
> Since `Collection` extends `Iterable` I understand why this is happening just not sure what the recommended action would be...
> any ideas/help would be appreciated :) I can push the code up to a branch if it's easier.

I think the easiest way for those tests is to declare a local variable, e.g.:

```java
  @Test
  void should_create_Assert() {
    Iterable<Object> actual = newLinkedHashSet();
    AbstractIterableAssert<?, Iterable<? extends Object>, Object, ObjectAssert<Object>> assertThat = Assertions.assertThat(actual);
    assertThat(assertThat).isNotNull();
  }
```

**Comment by HaydenMeloche** [comment]
@scordio I was thinking I had set something up incorrectly somewhere else. Casting it to an `Iterable<>` works for me :)
Thanks!

**Comment by HaydenMeloche** [comment]
@scordio When you have some time do you mind taking another look? I'm running into an issue where objects that extend `Collection` seem to not be able to find existing asserts... Not sure exactly what is causing the problem.

**Comment by scordio** [comment]
@HaydenMeloche I imagine the issue happens only with lists, right? I tried the following on top of your branch:

```java
    assertThat(Collections.singleton("element")) // CollectionAssert
      .isUnmodifiable(); // exists
    assertThat(Collections.singletonList("element")) // ListAssert
      .isUnmodifiable(); // does not exist
```

Probably making `AbstractListAssert` extend from `AbstractCollectionAssert` instead of `AbstractIterableAssert` should solve it, without introducing breaking changes.

Later on, you might need to add a `ClassBasedNavigableCollectionAssert` and a `ProxyableCollectionAssert` (you could take inspiration from the corresponding iterable ones).

It might help to visualize the class hierarchy for this topic:

class-diagram

File: src/main/java/org/assertj/core/internal/Collection.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import org.assertj.core.api.AssertionInfo;
+import org.assertj.core.util.VisibleForTesting;
+
+import static org.assertj.core.api.Fail.failBecauseExceptionWasNotThrown;
+
+/**
+ * Reusable assertions for <code>{@link java.util.Collection}</code>
+ */
+public class Collection {
```
Comment (scordio) [inline src/main/java/org/assertj/core/internal/Collection.java:23]: To follow the naming convention of similar classes, this should be `Collections`

**Comment by scordio** [comment]
Hi @HaydenMeloche, I have only now noticed that the PR has been closed for several weeks. It is absolutely fine if you don't want to continue or don't have time to do so, but we are happy to help if you would like to return to this or any other topic at a later time. 

As we will participate in the upcoming hack-commit-push event, I will unassign the issue for now and mark it as a possible task for the event.

**Pull request closed without merging**